File: leetcode/implementation/892. Surface Area of 3D Shapes.cpp
class Solution {
public:
    int dr[4] = {0,0,1,-1};
    int dc[4] = {1,-1,0,0};
    int surfaceArea(vector<vector<int>>& grid) {
        int ans = 0;
        int height = grid.size();
        int width = grid[0].size();
        for(int r = 0; r < height; r++) {
            for(int c = 0; c < width; c++) {
                int cnt = 0;
                for(int dir = 0; dir < 4; dir++) {
                    int nr = r + dr[dir];
                    int nc = c + dc[dir];
                    if(0 > nr || nr >= height || 0 > nc || nc >= width) {
                        ans += grid[r][c]; continue; //가장자리면
                    }
                    if(grid[r][c] > grid[nr][nc])
                        ans += grid[r][c] - grid[nr][nc]; //옆면
                }
                if(grid[r][c]) ans += 2; //윗면 아랫면
            }
        }
        return ans;
    }
};

File: leetcode/implementation/892. Surface Area of 3D Shapes/892. Surface Area of 3D Shapes.rs
impl Solution {
    pub fn surface_area(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let mut ans = 0;
        let dr: [i32; 4] = [0, 0, 1, -1];
        let dc: [i32; 4] = [1, -1, 0, 0];
        for r in 0..n {
            for c in 0..n {
                for dir in 0..4 {
                    let nr = r as i32 + dr[dir];
                    let nc = c as i32 + dc[dir];
                    if nr < 0 || nr >= n as i32 || nc < 0 || nc >= n as i32 {
                        ans += grid[r][c];
                        continue;
                    }
                    if grid[r][c] > grid[nr as usize][nc as usize] {
                        ans += grid[r][c] - grid[nr as usize][nc as usize];
                    }
                }
                if grid[r][c] > 0 {
                    ans += 2;
                }
            }
        }
        return ans;
    }
}
